docs/proposals/201909_thanos_sharding.md

```diff
@@ -0,0 +1,178 @@
+    
+## No Goals
+
+* Time partitioning
+* "Merging" sources together virtually for downsampling/compactions across single Source that just changes external labels.
+* Bloom filters (or custom advertised labels) for app metrics within blocks.
+* Allow all permutations of object storage setups:
+    * User uses multiple object storages for all sources?
+    * User uses single object storage for all sources?
+    * User uses any mix of object storages for sources. They put all in different subdirs/subpaths.
+* Add coordination or reconciliation in case of multi Compactor run on the same "sources" or any form of Compactor HA (e.g active passive)
+    * Requires separate design.
+
+This design is orthogonal for multi-buckets/multi-prefix support. 
+
+## Proposal
+
+We identified two use cases for relabeling on Thanos components:
+
+* Sharding: Operating on portion of object storage.
+* Advertising: Advertising custom labels on Info. This will result in pre-filtering on Querier which will not touch the server
+
+### Sharding
+
+On each component that works on the object storage (e.g Store GW and Compactor), add `--selector.relabel-config` (and corresponding `--selector.relabel-config-file`) that will
+be used to filter out what blocks should be selected for operations. Examples:
+
+* We want to run Compactor only for blocks with `external_labels` being `cluster=A`. We will run second Compactor for blocks with `cluster=B` external labels.
+* We want to browse only blocks with `external_labels` being `cluster=A` from object storage. We will run StoreGateway with selector of `cluster=A` from external labels of blocks.
+
+### Advertising
+
+On each component that exposes StoreAPI (e.g Querier, Ruler, Receiver, Store GW), add `--selector.relabel-config` (and corresponding `--selector.relabel-config-file`) that will
+be used to filter out what blocks should be selected for operations. Examples:
+
+* We want to run Ruler that will produce blocks with `external_labels` being `ruler_cluster="A", ruler_replica="1"`. This will ensure only those 
+
+NOTE: `selector.*` means that for those two use cases (Adv vs Sharding) we will use the same flags. This makes sense as essentially those are all about `selecting` what subset of metrics is given component responsible for.
+
+### Relabelling
+
+Similar to [promtail](https://github.com/grafana/loki/blob/master/docs/promtail.md#scrape-configs) this config will follow native
+[Prometheus relabel-config](https://prometheus.io/docs/prometheus/latest/configuration/configuration/#relabel_config) syntax.
+
+The relabel config will define filtering process done on **every** synchronization with object storage.
+
+We will allow manipulating with several of inputs:
+
+* `__block_id`
+* External labels:
+  * `<name>` 
+* `__block_objstore_bucket_endpoint` 
+* `__block_objstore_bucket_name`
+* `__block_objstore_bucket_path`
```

Comment:
This is a bit confusing to me. So let's say there will be a special `__block_objstore_bucket_path` label,  how can we manipulate it?

As I understand it we will be able to configure Thanos Store gateway to filter out folders from s3?

``` 
  action: drop
  regex: "folder_name"
  source_labels:
  - __block_objstore_bucket_path
```

Not sure how we can we do with  `__block_objstore_bucket_endpoint`, `__block_objstore_bucket_name` as they will be constant on the time?

Reply:
Yes, for now the options are limited - maybe indeed we should limit this (:

Reply:
We can talk about those details on the PR I guess.